specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,175 @@
+| **Attribute Key** | **Attribute Value** | **Type**       |
+|-------------------|---------------------|----------------|
+|  `foo.bar`        | `baz`               | `string`       |
+|  `my_array`       | `[first, second]`   | `string array` |
+|  `testnum*`       | `42.42`             | `number`       |
+
+the following labels must be used for the resulting metricset:
+
+| **Label Key** | **Label Value** | **Type**       |
+|---------------|-----------------|----------------|
+|  `foo_bar`    | `baz`           | `string`       |
+|  `testnum_`   | `42.42`         | `number`       |
+
+The OpenTelemetry specification allows the definition of metrics with the same name as long as they reside within a different instrumentation scope ([spec link](https://opentelemetry.io/docs/reference/specification/metrics/api/#get-a-meter)). In order to avoid conflicts, agents MUST add a mandatory label to the metricset with the key `otel_instrumentation_scope_name` and the corresponding instrumentation scope name as label value.
+
+## Exporter Installation
+
+The onboarding should be as easy as possible for users. Ideally, the users should be able to browse their OpenTelemetry metrics by just following the standard agent installation process without requiring additional configuration or code changes.
+
+We focus on two ways of how OpenTelemetry may be used by users:
+
+1. The user application has a OpenTelemetry Metrics SDK Instance setup and configured. E.g. the user already has a prometheus exporter running and did some customizations on their metrics via [Views](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#view).
+2. The user application does not setup a OpenTelemetry Metrics SDK Instance, but uses the OpenTelemetry Metrics API to define metrics. This is usually the case when the user application does not care / know about OpenTelemetry metrics, but a library used by the application implemented observability via OpenTelemetry.
+
+For **1.** we want to make sure that the existing user configuration is respected by our apm agents. Ideally, agents SHOULD just register an additional exporter to the existing OpenTelemetry Metrics SDK instance(s). If the agent and language capabilities allow it, the exporter registration SHOULD be done automatically without requiring code changes by the user. 
+
+For **2.** agents MAY automatically register an agent-provided SDK instance to bind the user provided OpenTelemetry API to, if this is possible in their language and does not cause too much overhead of any kind (e.g. implementation or agent package size).
+Agents MUST NOT override a user-provided global OpenTelemetry metrics SDK with their own SDK.
+For example the Java agent MUST NOT install an OpenTelemetry Metrics SDK instance in the [GlobalOpenTelemetry](https://www.javadoc.io/static/io.opentelemetry/opentelemetry-api/1.20.0/io/opentelemetry/api/GlobalOpenTelemetry.html) if it detects that another Metrics SDK has already been registered there.
```

Comment:
Is this sufficient? What if user code *later* attempts to register a `GlobalOpenTelemetry`?

Let me give a Node.js example where the handling of registering global OTel things seems to be slightly different. In the OTel JS code there is a `@opentelemetry/api` package with which setup code can register a global provider for tracing:

```js
const api = require('@opentelemetry/api')
api.trace.setGlobalTracerProvider(...)
```

or a "meter provider":

```js
const api = require('@opentelemetry/api')
api.metrics.setGlobalMeterProvider(...)
```

etc.

If JS code attempts call `setGlobalMeterProvider(...)` a *second* time, one gets this error message and the new provider does *not* get registered:

```
Error: @opentelemetry/api: Attempted duplicate registration of API: metrics
    at registerGlobal (/Users/trentm/el/apm-agent-nodejs17/examples/opentelemetry-metrics/node_modules/@opentelemetry/api/build/src/internal/global-utils.js:32:21)
    at MetricsAPI.setGlobalMeterProvider (/Users/trentm/el/apm-agent-nodejs17/examples/opentelemetry-metrics/node_modules/@opentelemetry/api/build/src/api/metrics.js:41:50)
    at Object.<anonymous> (/Users/trentm/el/apm-agent-nodejs17/examples/opentelemetry-metrics/otel-prom.js:30:14)
    at Module._compile (node:internal/modules/cjs/loader:1165:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1219:10)
    at Module.load (node:internal/modules/cjs/loader:1043:32)
    at Function.Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:47
```

When using the Node.js APM agent the **goal** is that users start the agent before most anything else is done in the application. So the typical case is that the Node.js APM agent can't know if user code will later attempt to `.setGlobalMeterProvider(...)`.

I guess all I'm saying is:
- We could consider a dev/implementation note here for usage of the OTel JS SDK that the APM agent cannot use the straightforward `api.metrics.setGlobalMeterProvider(...)` option.
- Does the Java APM agent have a similar concern with *later* user code attempting to use [`GlobalOpenTelemetry.set(...)`](https://www.javadoc.io/static/io.opentelemetry/opentelemetry-api/1.20.0/io/opentelemetry/api/GlobalOpenTelemetry.html#set(io.opentelemetry.api.OpenTelemetry))?

Reply:
In java the following rules apply to `GlobalOpenTelemetry`:
 * `GlobalOpenTelemetry.set` may be invoked at most once
 * `GlobalOpenTelemetry.set` may not be invoked after `GlobalOpenTelemetry.get` has been called

In the Java agent we don't actually call `GlobalOpenTelemetry.set` ever. Instead, we instrument `GlobalOpenTelemetry.get` and replace the return value with our own implementation if required.
Because `GlobalOpenTelemetry.set` may not be called anymore after `GlobalOpenTelemetry.get` is called, we can be sure that if the user wanted to configure his own SDK instance, it has already happened.

I don't know if the same approach can be applied to Node.js, but I agree that the spec might be a bit ambiguous here. The goal should be rather: `If the user wants to provide his own SDK instance, the agent MUST NOT prevent him from doing so`. So if the agent is capable of detecting this automatically (like Java): Nice! If not: the automatic installation of the agent's metrics SDK instance should be opt-in via configuration.

Reply:
Understood. Thanks.

For the OTel *Bridge*, the Node.js APM agent currently *is* setting a global trace provider: https://github.com/elastic/apm-agent-nodejs/blob/v3.42.0/lib/opentelemetry-bridge/setup.js#L53-L67
The OTel Bridge is off by default, and enabled via `opentelemetry_bridge_enabled=true` so I think this is fine for now.

For the OTel Metrics work, I'll have to do the equivalent thing to what the Java agent is doing: instrument `.metrics.getMeterProvider()` method from the `@opentelemetry/api` package.